exporter.go

```diff
@@ -639,17 +641,67 @@ samples:
 }
 
 type StatsDUDPListener struct {
-	conn *net.UDPConn
+	conn          *net.UDPConn
+	bufferWatcher *telemetry.BufferWatcher
+	bufferWarning uint32
 }
 
-func (l *StatsDUDPListener) Listen(e chan<- Events) {
+func (l *StatsDUDPListener) Listen(threadCount int, packetHandlers int, e chan<- Events) {
+	bufferWatcher, err := telemetry.NewBufferWatcher(l.conn)
+	l.bufferWatcher = bufferWatcher
+	shouldWatchBuffer := true
+
+	if err != nil {
+		shouldWatchBuffer = false
+		log.Debugf("Unable to watch UDP buffer size due to: %v", err)
+	}
+
+	l.bufferWarning = 0
+
+	concurrentHandlersPerThread := packetHandlers / threadCount
+
+	for i := 0; i < threadCount; i++ {
+		go l.Listener(e, concurrentHandlersPerThread, shouldWatchBuffer)
+	}
+}
+
+func (l *StatsDUDPListener) Listener(e chan<- Events, concurrentPacketHandlers int, shouldWatchBuffer bool) {
+	var sem = make(chan struct{}, concurrentPacketHandlers)
 	buf := make([]byte, 65535)
 	for {
 		n, _, err := l.conn.ReadFromUDP(buf)
 		if err != nil {
 			log.Fatal(err)
 		}
-		l.handlePacket(buf[0:n], e)
+		if shouldWatchBuffer && atomic.LoadUint32(&l.bufferWarning) != 1 {
+			go l.watchBufferSize()
+		}
+
+		data := append([]byte(nil), buf[0:n]...)
+		select {
+		case sem <- struct{}{}:
+			{
+				go func() {
+					l.handlePacket(data[0:n], e)
```

Comment:
is it safe to call handlePacket concurrently?

why a semaphore and goroutines per `handlePacket` call, vs long-lived goroutines that consume off a (buffered) channel? IIUC, with the buffered channel it would be possible to try to write an event into this channel, but if the buffer is full, increment a drop counter and move on?

Reply:
It is safe! At least, I ran this with a `-race` build with no complaints, and have ~22 deployments of this happily chugging along at a pretty high volume and no related crashes.

Re: Semaphore - I think that is a great suggestion, this was originally @AustinSchoen's work, lemme collab with him. The idea of reading UDP packets as fast as we can and dropping them if the processing can't keep up instead of blocking (but keeping track of that drop tally) is so elegant, very +++ to headed down that path. That _should_ totally alleviate my need for procfs metrics.

Reply:
all credit for that idea goes to @rtreffer!

Reply:
I believe I went with a semaphore because it was the first example of a tunable queue of workers that I could find, there are multiple ways of handling this model. We were still early in the experimentation phase as to what amount and kind of concurrency we needed to handle the firehose of UDP packets coming at our exporters. Regardless I don't think we'd want to just drop data on the floor if we don't have the capacity to handle it in every situation. Those packets should back up into the protocol buffer while waiting to be read by the configured number of handlers.

In situations where a system is overloaded only for a few (mili)seconds it would be better to block read off the protocol buffer and wait to handle them than to drop and log that we lost packets IMO.

Reply:
> I don't think we'd want to just drop data on the floor if we don't have the capacity to handle it in every situation

I agree, and I don't propose that we do that. We can buffer again after the initial receive.

However, at some point we _have_ to drop, and we want to know when that happens. I would rather guarantee that that happens in a place where we can easily observe it (i.e. in user-lang Go code) than introduce complex platform-dependent monitoring of the socket buffers themselves.

Reply:
With "buffer again" I mean stuff them into a buffered channel, we don't have to get fancy about it.